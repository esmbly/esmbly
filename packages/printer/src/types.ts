export type Callback = (error: Error | null | undefined) => void;

export type Encoding = string | undefined;

export type WriteFn = (
  chunk: string,
  encoding?: Encoding,
  callback?: Callback,
) => boolean;
