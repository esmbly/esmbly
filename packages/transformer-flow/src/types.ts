import t from '@babel/types';
import { Node, Visitor } from '@babel/traverse';

export interface Warning {
  info: string;
  issueUrl: string;
  node: t.Node;
}

type RuleA = () => Visitor<Node>;
type RuleB = (warnings: Warning[]) => Visitor<Node>;
export type Rule = RuleA | RuleB;
