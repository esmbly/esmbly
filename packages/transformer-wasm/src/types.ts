import t from '@babel/types';
import { Node, Visitor } from '@babel/traverse';

export interface Warning {
  info: string;
  issueUrl?: string;
  node: t.Node;
}

type RuleA = () => Visitor<Node>;
type RuleB = (warnings: Warning[]) => Visitor<Node>;
export type Rule = RuleA | RuleB;

export interface WasmTransformerOptions {
  optimize?: string;
  optimizeLevel?: number;
  shrinkLevel?: number;
  validate?: boolean;
  // TODO: ADD more options
}
