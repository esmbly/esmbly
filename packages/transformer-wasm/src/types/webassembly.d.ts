/*
WebAssembly v1 (MVP) declaration file for TypeScript
Definitions by: 01alchemist (https://twitter.com/01alchemist)

MIT License

Copyright (c) 2017 01Alchemist

Permission is hereby granted, free of charge, to any person obtaining a copy
of this software and associated documentation files (the "Software"), to deal
in the Software without restriction, including without limitation the rights
to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
copies of the Software, and to permit persons to whom the Software is
furnished to do so, subject to the following conditions:

The above copyright notice and this permission notice shall be included in all
copies or substantial portions of the Software.

THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
SOFTWARE.
*/

declare namespace WebAssembly {
  /**
   * WebAssembly.Module
   **/
  class Module {
      constructor (bufferSource: ArrayBuffer | Uint8Array);

      static customSections(module: Module, sectionName: string): ArrayBuffer[];
      static exports(module: Module): {
          name: string;
          kind: string;
      }[];
      static imports(module: Module): {
          module: string;
          name: string;
          kind: string;
      }[];
  }

  /**
   * WebAssembly.Instance
   **/
  class Instance {
      readonly exports: any;
      constructor (module: Module, importObject?: any);
  }

  /**
   * WebAssembly.Memory
   * Note: A WebAssembly page has a constant size of 65,536 bytes, i.e., 64KiB.
   **/
  interface MemoryDescriptor {
      initial: number;
      maximum?: number;
  }

  class Memory {
      readonly buffer: ArrayBuffer;
      constructor (memoryDescriptor: MemoryDescriptor);
      grow(numPages: number): number;
  }

  /**
   * WebAssembly.Table
   **/
  interface TableDescriptor {
      element: "anyfunc",
      initial: number;
      maximum?: number;
  }

  class Table {
      readonly length: number;
      constructor (tableDescriptor: TableDescriptor);
      get(index: number): Function;
      grow(numElements: number): number;
      set(index: number, value: Function): void;
  }

  /**
   * Errors
   */
  class CompileError extends Error {
      readonly fileName: string;
      readonly lineNumber: string;
      readonly columnNumber: string;
      constructor (message?:string, fileName?:string, lineNumber?:number);
      toString(): string;
  }

  class LinkError extends Error {
      readonly fileName: string;
      readonly lineNumber: string;
      readonly columnNumber: string;
      constructor (message?:string, fileName?:string, lineNumber?:number);
      toString(): string;
  }

  class RuntimeError extends Error {
      readonly fileName: string;
      readonly lineNumber: string;
      readonly columnNumber: string;
      constructor (message?:string, fileName?:string, lineNumber?:number);
      toString(): string;
  }

  function compile(bufferSource: ArrayBuffer | Uint8Array): Promise<Module>;

  interface ResultObject {
      module: Module;
      instance: Instance;
  }

  function instantiate(bufferSource: ArrayBuffer | Uint8Array, importObject?: any): Promise<ResultObject>;
  function instantiate(module: Module, importObject?: any): Promise<Instance>;

  function validate(bufferSource: ArrayBuffer | Uint8Array): boolean;
}