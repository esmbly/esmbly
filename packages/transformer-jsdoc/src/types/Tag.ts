export interface TagType {
  type: string;
  name: string;
  elements?: TagType[];
  expression: TagType;
  applications: TagType[];
}

export interface Tag {
  title: string;
  description: string;
  type: TagType;
  name: string;
}
