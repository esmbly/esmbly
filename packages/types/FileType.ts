export enum FileType {
  JavaScript = 'JavaScript',
  Flow = 'Flow',
  TypeScript = 'TypeScript',
  WebAssembly = 'WebAssembly',
}
