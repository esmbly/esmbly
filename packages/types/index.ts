export * from './AST';
export * from './File';
export * from './FileType';
export * from './RunConfig';
export * from './Transformer';
