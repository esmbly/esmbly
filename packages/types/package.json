{
  "name": "@esmbly/types",
  "version": "0.0.6",
  "description": "Type definitions for Esmbly, a composable transpiler infrastructure for working with statically typed JavaScript and WebAssembly",
  "author": "Olle Lauri Bostr√∂m <ollebostr@gmail.com> (https://github.com/ollelauribostrom)",
  "homepage": "https://github.com/esmbly/esmbly",
  "license": "MIT",
  "main": "dist/index.js",
  "types": "dist/index.d.ts",
  "repository": {
    "type": "git",
    "url": "git+https://github.com/esmbly/esmbly.git"
  },
  "bugs": {
    "url": "https://github.com/esmbly/esmbly/issues"
  },
  "scripts": {
    "build": "tsc",
    "build-watch": "tsc -w"
  },
  "dependencies": {
    "@babel/parser": "^7.4.3",
    "@babel/traverse": "^7.4.3",
    "@babel/types": "^7.3.4"
  },
  "publishConfig": {
    "access": "public"
  }
}
