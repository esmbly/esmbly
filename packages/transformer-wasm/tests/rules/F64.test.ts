import { Format } from '@esmbly/types';
import testRunner from '../helpers/testRunner';

const program = `
  function add(a: number, b: number): number {
    return a + b;
  }
`;

describe('rule: F64', () => {
  it('transforms number -> f64', async () => {
    const output = [{ format: Format.AssemblyScript }];
    const [{ content }] = await testRunner(program, output);
    expect(content).toMatchSnapshot();
  });
});
