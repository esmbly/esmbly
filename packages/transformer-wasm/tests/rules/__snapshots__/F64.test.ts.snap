// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`rule: F64 transforms number -> f64 1`] = `
"
  function add(a: f64, b: f64): f64 {
    return a + b;
  }
"
`;
